source/common/upstream/original_dst_cluster.cc

```diff
@@ -166,48 +167,82 @@ OriginalDstCluster::OriginalDstCluster(
-  auto pair = new_host_map->emplace(host->address()->asString(), host);
-  bool added = pair.second;
-  if (added) {
-    ENVOY_LOG(debug, "addHost() adding {}", host->address()->asString());
-    setHostMap(new_host_map);
-    // Given the current config, only EDS clusters support multiple priorities.
-    ASSERT(priority_set_.hostSetsPerPriority().size() == 1);
-    const auto& first_host_set = priority_set_.getOrCreateHostSet(0);
-    HostVectorSharedPtr all_hosts(new HostVector(first_host_set.hosts()));
-    all_hosts->emplace_back(host);
-    priority_set_.updateHosts(0,
-                              HostSetImpl::partitionHosts(all_hosts, HostsPerLocalityImpl::empty()),
-                              {}, {std::move(host)}, {}, absl::nullopt);
+  std::string address = host->address()->asString();
+  HostMultiMapSharedPtr new_host_map = std::make_shared<HostMultiMap>(*getCurrentHostMap());
+  auto it = new_host_map->find(address);
+  if (it != new_host_map->end()) {
+    it->second->hosts_.push_back(host);
+  } else {
+    new_host_map->emplace(address, std::make_shared<HostsForAddress>(host));
   }
+  ENVOY_LOG(debug, "addHost() adding {} {}.", *host, address);
+  setHostMap(new_host_map);
+
+  // Given the current config, only EDS clusters support multiple priorities.
+  ASSERT(priority_set_.hostSetsPerPriority().size() == 1);
+  const auto& first_host_set = priority_set_.getOrCreateHostSet(0);
+  HostVectorSharedPtr all_hosts(new HostVector(first_host_set.hosts()));
+  all_hosts->emplace_back(host);
+  priority_set_.updateHosts(0,
+                            HostSetImpl::partitionHosts(all_hosts, HostsPerLocalityImpl::empty()),
+                            {}, {std::move(host)}, {}, absl::nullopt);
 }
 
 void OriginalDstCluster::cleanup() {
   HostVectorSharedPtr keeping_hosts(new HostVector);
   HostVector to_be_removed;
-  ENVOY_LOG(trace, "Stale original dst hosts cleanup triggered.");
+  absl::flat_hash_set<absl::string_view> removed_addresses;
   auto host_map = getCurrentHostMap();
   if (!host_map->empty()) {
     ENVOY_LOG(trace, "Cleaning up stale original dst hosts.");
-    for (const auto& pair : *host_map) {
-      const std::string& addr = pair.first;
-      const HostSharedPtr& host = pair.second;
-      if (host->used()) {
-        ENVOY_LOG(trace, "Keeping active host {}.", addr);
-        keeping_hosts->emplace_back(host);
-        host->used(false); // Mark to be removed during the next round.
+    for (const auto& [addr, hosts] : *host_map) {
+      // Address is kept in the cluster if either of the two things happen:
+      // 1) a host has been recently selected for the address; 2) none of the
+      // hosts are currently in any of the connection pools.
+      // The set of hosts for a single address are treated as a unit.
+      //
+      // Using the used_ bit is preserved for backwards compatibility and to
+      // add a delay between load balancers choosing a host and grabbing a
+      // handle on the host.
```

Comment:
Can you spell out the interleaving that you put in the comment thread here? Basically that the used_ bit "gaurantees" that for at least one cleanup interval the host will remain. Obviously there is no real hard guarantee, but it's very likely given the time values involved.

Reply:
Ack.